src/graphql/queries.js

```diff
@@ -81,6 +86,116 @@ subscription {
 }
 `
 
+export const WORKFLOW_TREE_SUBSCRIPTION = gql`
+subscription ($workflowId: ID) {
+  deltas (workflows: [$workflowId]) {
+    id
+    shutdown
+    added {
+      workflow {
+        ...WorkflowData
+        cyclePoints: familyProxies(ids: ["root"], ghosts: true) {
+          cyclePoint
+        }
+        taskProxies(sort: { keys: ["cyclePoint"] }, ghosts: true) {
+          ...TaskProxyData
+          jobs(sort: { keys: ["submit_num"], reverse:true }) {
+            ...JobData
+          }
+        }
+        familyProxies (exids: ["root"], sort: { keys: ["firstParent"]}, ghosts: true) {
+          ...FamilyProxyData
+        }
+      }
+      cyclePoints: familyProxies(ids: ["root"], ghosts: true) {
+        cyclePoint
+      }
+      familyProxies (exids: ["root"], sort: { keys: ["firstParent"]}, ghosts: true) {
+        ...FamilyProxyData
+      }
+      taskProxies(sort: { keys: ["cyclePoint"] }, ghosts: true) {
+        ...TaskProxyData
+      }
+      jobs(sort: { keys: ["submit_num"], reverse:true }) {
+        ...JobData
+      }
+    }
+    updated {
+      taskProxies(sort: { keys: ["cyclePoint"] }, ghosts: true) {
+        ...TaskProxyData
+      }
+      jobs(sort: { keys: ["submit_num"], reverse:true }) {
+        ...JobData
+      }
+      familyProxies (exids: ["root"], sort: { keys: ["firstParent"]}, ghosts: true) {
+        ...FamilyProxyData
+      }
+    }
+    pruned {
+      jobs
+      taskProxies
+      familyProxies
+    }
+  }
+}
+
+fragment WorkflowData on Workflow {
```

Comment:
Nice way to avoid writing this out twice.

Could we use the above template for all subscriptions by just varying the fragments or does GraphQL not allow you to have empty fragments?

Reply:
We could use string-replacement to build up the query+fragments. That's what I am working now (more thinking and reading than coding), for https://github.com/cylc/cylc-ui/issues/439